File: frontend-angular/src/app/services/cliente.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Cliente } from '../models/cliente.model';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class ClienteService {
  private readonly apiUrl = `${environment.apiUrl}/clientes`;

  constructor(private http: HttpClient) {}

  listarClientes(): Observable<Cliente[]> {
    return this.http.get<Cliente[]>(this.apiUrl);
  }

  listarActivos(): Observable<Cliente[]> {
    return this.http.get<Cliente[]>(`${this.apiUrl}/activos`);
  }

  obtenerPorId(id: number): Observable<Cliente> {
    return this.http.get<Cliente>(`${this.apiUrl}/${id}`);
  }

  registrarCliente(cliente: Omit<Cliente, 'idCliente' | 'activo'>): Observable<Cliente> {
    return this.http.post<Cliente>(this.apiUrl, cliente);
  }

  actualizar(cliente: Cliente): Observable<Cliente> {
    return this.http.put<Cliente>(`${this.apiUrl}/${cliente.idCliente}`, cliente);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  desactivar(id: number): Observable<Cliente> {
    return this.http.patch<Cliente>(`${this.apiUrl}/${id}/desactivar`, {});
  }

  activar(id: number): Observable<Cliente> {
    return this.http.patch<Cliente>(`${this.apiUrl}/${id}/activar`, {});
  }

  buscarPorDni(dni: string): Observable<Cliente> {
    return this.http.get<Cliente>(`${this.apiUrl}/buscar`, { params: { dni } });
  }
}

File: ms-clientes/src/main/java/com/ferreteria/clientes/service/ClienteService.java
package com.ferreteria.clientes.service;

import com.ferreteria.clientes.dto.ClienteRequest;
import com.ferreteria.clientes.dto.ClienteResponse;
import java.util.List;

public interface ClienteService {
    List<ClienteResponse> listarClientes();
    List<ClienteResponse> listarActivos();
    ClienteResponse obtenerPorId(Long id);
    ClienteResponse buscarPorDni(String dni);
    ClienteResponse registrarCliente(ClienteRequest request);
    ClienteResponse actualizar(Long id, ClienteRequest request);
    void eliminar(Long id);
    ClienteResponse activar(Long id);
    ClienteResponse desactivar(Long id);
}
